### PR TITLE
CORE-179: create Match type class instance for rholang

### PR DESCRIPTION
**Pull request by henrytill**
Depends on #388.

https://rchain.atlassian.net/browse/CORE-179

By making a minor modification to @guardbotmk3's `foldMatch` from the above PR, I was basically able to leverage the Rholang matcher code directly, without any need for code duplication.

In order to avoid to avoid a circular dependency between `rholang` and `storage`, I moved the `Serialize` type class definition back to `storage` (where it belongs, along with the `Match` type class, IMHO).  

I also think we might considering moving the code from `models` into the `rholang` project and doing away with that subproject.  Curious if anyone else has thoughts on this...

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+          case ((i1, _), (i2, _)) => i1.compare(i2)
+        }
+    }
+
+  private def toQuotes[T](fm: FreeMap)(implicit ord: Ordering[(Int, Par)]): Option[List[Quote]] = {
+    @tailrec
+    def loop(sortedParList: List[(Int, Par)], curr: Int, acc: List[Quote]): Option[List[Quote]] =
+      sortedParList match {
+        case Nil                         => Some(acc)
+        case (i, t) :: tail if i == curr => loop(tail, i + 1, Quote(t) :: acc)
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: What is the purpose of tracking the `i`? Don't we in the end just get rid of that information and just return the List of Quotes?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala [hunk 2]
```diff
@@ -65,14 +65,14 @@ object SpatialMatcher {
       .reverse
 
   // This helper function is useful in several productions
-  def foldMatch(tlist: Seq[Par], plist: Seq[Par]): OptionalFreeMap[Unit] =
+  def foldMatch[T](tlist: Seq[T],
+                   plist: Seq[T],
+                   f: (T, T) => OptionalFreeMap[Unit]): OptionalFreeMap[Unit] =
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala]: Why is this necessary? Does scala not know how to dispatch to the corresponding methods

spatialMatch(target: Par, pattern: Par)
spatialMatch(target: Channel, pattern: Channel)

based on the types passed in?

Comment (henrytill) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala]: `T` is too generic for that.  There isn't a `spatialMatch` for any old type, so taking that approach, you'd have to constrain `T` with a type class (something like `trait SpatialMatch[A]`).  Then you'd have in something like "any `T` that has an instance of `SpatialMatch`" (aka something of type `SpatialMatch[T]`).

Or alternately, if `Par` or `Channel` were both subtypes of another type (or even if one was a subtype of the other), we could use the supertype and we could use `spatialMatch` in the body of the function.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala (hunk 1 above)
Comment (henrytill) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: I'm curious about the why too. 

@guardbotmk3?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala (hunk 2 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala]: I see. I guess in the future everything as in (Par, Expr, Send, Receive, etc) should extend from a trait Term and then have a generic spatialMatch(target: Term, pattern: Term). But yeah I guess the other spatialMatch methods need to be implemented first.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala (hunk 1 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: So I can see that it is useful to return from the matcher if there was another Tuplespace implementation that didn't assume the order that the quotes were returned. But since the Tuplespace implementation you wrote just assumes to the order, I think we can just immediately drop it as in

def loop(sortedParList: List[(Int, Par)]) = sortedParList.map(Quote(_._2))

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: Hmm actually we need to know how to then substitute.

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: Are you sure we can just return Option[List[Quote]] for 

def get(patterns: List[Channel], data: List[Quote]): Option[List[Quote]]

. I think it might be handy to pass through the level info back to the interpreter.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import coop.rchain.models.Channel.ChannelInstance.Quote
+import coop.rchain.storage.{Match => StorageMatch}
+import coop.rchain.rholang.interpreter.SpatialMatcher._
+
+import scala.annotation.tailrec
+
+//noinspection ConvertExpressionToSAM
+object implicits {
+
+  implicit def orderingIntTuple[T]: Ordering[(Int, T)] =
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: You can use the `Ordering.by` method:
```
def orderingIntTuple[T]: Ordering[(Int, T)] = Ordering.by[(Int, T), Int](_._1)

```

Comment (henrytill) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: Thanks for pointing that out!  I had forgotten about `Ordering.by`...

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala (hunk 2 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala]: I'm fine with this change and the associated changes here.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala (hunk 1 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: The point is that each free variable in the pattern needs an entry in the list.
Unfortunately freeCount isn't exactly what you want, since I rolled wildcards into it.
Can you @henrytill or @KentShikama write a small PR that splits out freecount from a wildcard boolean.

Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: patterns.map(freeCount).sum

**Comment by henrytill** [comment]
@guardbotmk3 @KentShikama - ready for re-review.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import coop.rchain.rholang.interpreter.SpatialMatcher._
+import coop.rchain.rholang.interpreter.implicits._
+import coop.rchain.storage.{Match => StorageMatch}
+
+import scala.collection.immutable.SortedMap
+
+//noinspection ConvertExpressionToSAM
+object implicits {
+
+  private def toQuotes[T](fm: FreeMap, max: Int): List[Quote] = {
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: Unnecessary [T]

Comment (henrytill) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/implicits.scala]: whoop, thanks
